core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/AdaptiveBroadcastToPartitionedJoin.java

```diff
@@ -0,0 +1,464 @@
+            this.globalContext = requireNonNull(globalContext, "globalContext is null");
+        }
+
+        @Override
+        protected PlanNode visitPlan(PlanNode node, RewriteContext<Void> context)
+        {
+            return partitionedExchange(
+                    globalContext.getIdAllocator().getNextId(),
+                    REMOTE,
+                    node,
+                    probeSymbols,
+                    Optional.empty());
+        }
+
+        @Override
+        public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> context)
+        {
+            if (node.getStep() == PARTIAL) {
+                return rewriteSources(this, node, globalContext);
+            }
+            return visitPlan(node, context);
+        }
+
+        @Override
+        public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
+        {
+            verify(node.getScope() == REMOTE && node.getType() == REPARTITION);
+            return partitionedExchange(
+                    globalContext.getIdAllocator().getNextId(),
+                    REMOTE,
+                    node.getSources().get(0),
+                    probeSymbols,
+                    Optional.empty());
+        }
+
+        @Override
+        public PlanNode visitRemoteSource(RemoteSourceNode node, RewriteContext<Void> context)
+        {
+            verify(node.getExchangeType() == REPARTITION);
+            return visitPlan(node, context);
+        }
+    }
+
+    private static class BuildSideRewriter
+            extends SimplePlanRewriter<Void>
+    {
+        private final List<Symbol> buildSymbols;
+        private final Context globalContext;
+
+        private BuildSideRewriter(List<Symbol> buildSymbols, Context globalContext)
+        {
+            this.buildSymbols = requireNonNull(buildSymbols, "buildSymbols is null");
+            this.globalContext = requireNonNull(globalContext, "globalContext is null");
+        }
+
+        @Override
+        protected PlanNode visitPlan(PlanNode node, RewriteContext<Void> context)
+        {
+            return partitionedExchange(
+                    globalContext.getIdAllocator().getNextId(),
+                    REMOTE,
+                    node,
+                    buildSymbols,
+                    Optional.empty());
+        }
+
+        @Override
+        public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> context)
+        {
+            if (node.getStep() == PARTIAL) {
+                return rewriteSources(this, node, globalContext);
+            }
+            return visitPlan(node, context);
+        }
+
+        @Override
+        public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
+        {
+            if (node.getScope() == LOCAL) {
+                return rewriteSources(this, node, globalContext);
+            }
+            verify(node.getScope() == REMOTE && node.getType() == REPLICATE);
```

Comment:
should you also verify that `node.partitioningColumns` were not set or that using `buildSymbols` as partitioning columns is compatible with we had previously (`buildSymbols` would need to be subset of previous set)